src/lib/datepicker/calendar.ts

```diff
@@ -97,57 +107,181 @@ export class MdCalendar implements AfterContentInit {
   }
 
   ngAfterContentInit() {
-    this._currentPeriod = this.startAt || SimpleDate.today();
+    this._activeDate = this.startAt || SimpleDate.today();
     this._monthView = this.startView != 'year';
   }
 
   /** Handles date selection in the month view. */
-  _dateSelected(date: SimpleDate) {
+  _dateSelected(date: SimpleDate): void {
     if ((!date || !this.selected) && date != this.selected || date.compare(this.selected)) {
       this.selectedChange.emit(date);
     }
   }
 
   /** Handles month selection in the year view. */
-  _monthSelected(month: SimpleDate) {
-    this._currentPeriod = month;
+  _monthSelected(month: SimpleDate): void {
+    this._activeDate = month;
     this._monthView = true;
   }
 
   /** Handles user clicks on the period label. */
-  _currentPeriodClicked() {
+  _currentPeriodClicked(): void {
     this._monthView = !this._monthView;
   }
 
   /** Handles user clicks on the previous button. */
-  _previousClicked() {
-    let amount = this._monthView ? {months: -1} : {years: -1};
-    this._currentPeriod = this._currentPeriod.add(amount);
+  _previousClicked(): void {
+    this._activeDate = this._monthView ?
+        this._addCalendarMonths(this._activeDate, -1) :
+        this._addCalendarYears(this._activeDate, -1);
   }
 
   /** Handles user clicks on the next button. */
-  _nextClicked() {
-    let amount = this._monthView ? {months: 1} : {years: 1};
-    this._currentPeriod = this._currentPeriod.add(amount);
+  _nextClicked(): void {
+    this._activeDate = this._monthView ?
+        this._addCalendarMonths(this._activeDate, 1) : this._addCalendarYears(this._activeDate, 1);
   }
 
   /** Whether the previous period button is enabled. */
-  _previousEnabled() {
+  _previousEnabled(): boolean {
     if (!this.minDate) {
       return true;
     }
-    return !this.minDate || !this._isSameView(this._currentPeriod, this.minDate);
+    return !this.minDate || !this._isSameView(this._activeDate, this.minDate);
   }
 
   /** Whether the next period button is enabled. */
-  _nextEnabled() {
-    return !this.maxDate || !this._isSameView(this._currentPeriod, this.maxDate);
+  _nextEnabled(): boolean {
+    return !this.maxDate || !this._isSameView(this._activeDate, this.maxDate);
   }
 
   /** Whether the two dates represent the same view in the current view mode (month or year). */
-  private _isSameView(date1: SimpleDate, date2: SimpleDate) {
+  private _isSameView(date1: SimpleDate, date2: SimpleDate): boolean {
     return this._monthView ?
         date1.year == date2.year && date1.month == date2.month :
         date1.year == date2.year;
   }
+
+  /** Handles keydown events on the calendar body. */
+  _handleCalendarBodyKeydown(event: KeyboardEvent): void {
```

Comment:
Considered breaking this logic out into its own mini-service? 

Also, have you thought about separating into two functions -one for month view and one for year view? Not sure if that will actually shorten anything, but it may be easier to scan it without the conditionals.